### DISCUSSION TITLE
Hydration Error with NextJS 15

### DISCUSSION THREAD

**Discussion by ivan-dalmet** [chakra-ui/chakra-ui#9051] [General]
### Description

Getting hydratation error even with `suppressHydratationWarning` in NextJS 15 (with React 19)
![image](https://github.com/user-attachments/assets/63f66869-5bad-47c3-bf4c-b28c5a03390f)

### Link to Reproduction

https://codesandbox.io/p/devbox/77hrfw

### Steps to reproduce

1. Open the preview
2. You will get an hydratation error

### Chakra UI Version

3.0.0-next.32 (but same error with 2.x)

### Browser

Ark

### Operating System

- [X] macOS
- [ ] Windows
- [ ] Linux

### Additional Information

Maybe the `suppressHydratationWarning` does not work anymore in NextJS 15? or React 19?

PS: Thanks for Charka UI and the new v3 ❤️

**Comment by lcswillems** [comment]
@segunadebayo In the doc, it says to add `suppressHydratationWarning` to hide the warning, but we should not hide the warning. If the warning is displayed, it is because Chakra 3 doesn't something incorrect, so instead Chakra 3 should be fixed?

**Comment by ivan-dalmet** [comment]
@lcswillems yes but also the issue here is that suppressHydratationWarning doesn't work 😟
So following the docs does not work

**Comment by ivan-dalmet** [comment]
By the way, the error doesn't seem to show at each reload 😥

**Reply by codetestmg** [reply]
yes exactly. even i am facing similar issue

**Comment by segunadebayo** [comment]
This is probably related to https://github.com/pacocoursey/next-themes or Next 15

@lcswillems here's what next-themes recommends. It's an intentional decision; the prop exists in react for edge cases like this.
![CleanShot 2024-10-25 at 17 32 06@2x](https://github.com/user-attachments/assets/aab5c94e-ac57-43b0-b0d6-78d47751273e)

**Reply by sayshark75** [reply]
Hi, i tried your solution on the issue #9576, but still not worked. Can you reopen the issue so i can get more solutions or workaround on the issue.

**Comment by baptisteArno** [comment]
Exact same issue here even with the `suppressHydrationWarning` workaround (React 18). Reverting to Next v14 fixes it

**Reply by oxedom** [reply]
That's the only solution that solved it for myself.

**Comment by unishigure** [comment]
Hi. I have the same problem in Next.js v15.0.1.
This problem only occurs when `--turbopack` option in my environment.

**Comment by PatrickRogg** [comment]
Adding an emotion cache solved it for me. From here: https://github.com/emotion-js/emotion/issues/2928#issuecomment-1293012737

```
'use client';

import { ChakraProvider } from '@chakra-ui/react';
import { ColorModeProvider } from '../components/ui/color-mode';
import { system } from '../style-system';
import { EmotionCacheProvider } from './emotion-cache';

export function DesignThemeProvider(props: { children: React.ReactNode }) {
  return (
    <EmotionCacheProvider>
      <ChakraProvider value={system}>
        <ColorModeProvider defaultTheme="light">{props.children}</ColorModeProvider>
      </ChakraProvider>
    </EmotionCacheProvider>
  );
}
```

```
'use client';

import React, { useState } from 'react';
import { useServerInsertedHTML } from 'next/navigation';
import { CacheProvider } from '@emotion/react';
import createCache from '@emotion/cache';

export function EmotionCacheProvider({ children }: { children: React.ReactNode }) {
  const [cache] = useState(() => {
    const cache = createCache({ key: 'chakra', prepend: true });
    // @ts-ignore
    cache.compat = true; // Ensure the cache is compatible with Emotion 10
    return cache;
  });

  useServerInsertedHTML(() => (
    <style
      data-emotion={`${cache.key} ${Object.keys(cache.inserted).join(' ')}`}
      dangerouslySetInnerHTML={{
        __html: Object.values(cache.inserted).join(' ')
      }}
    />
  ));

  return <CacheProvider value={cache}>{children}</CacheProvider>;
}
```

**Reply by PieroDev23** [reply]
bro this actually works for me, but why?

**Reply by hays0503** [reply]
@PatrickRogg Thanks for your solution! I found your answer after 19 hours of exhausting debugging, and I’m really glad it worked.

But I still don’t quite understand — why does it work?
Maybe you could write a post about it or suggest it to the project’s author?

**Reply by rowin1125** [reply]
I tried this solution and though it works, its very heavy on the dev console for some reason. This causes my browser to occasionally crash. Did anyone else also experience this?

**Reply by LDami** [reply]
Thanks for the solution ! Works for me on Next@16.0.5

**Comment by sayshark75** [comment]
Hey @segunadebayo , is this issue resolved? i didn't face any such issue of hydration, by following the app router instruction in docs.
also no hydration issue in the codesandbox link. or i missed something? Thanks.

**Comment by skylarmb** [comment]
I can confirm this only happens with `--turbopack`

**Comment by segunadebayo** [comment]
Given this issue only happens when using `--turbopack`, I'm not sure what we can do about it.

Moving it to a discussion until we have a definite plan of action.

**Comment by arealesramirez** [comment]
I'm currently experiencing a similar issue.

I'm using Next.js 15.0.2 and Chakra 3.1.0. I added the the `suppressHydrationWarning` flag, and it doesn't suprress the warning.

My `RootLayout` looks like this

```
import { draftMode } from "next/headers";
import { Inter } from "next/font/google";

import "@/app/globals.css";

import Navigation from "@/components/Globals/Navigation/Navigation";
import { PreviewNotice } from "@/components/Globals/PreviewNotice/PreviewNotice";
import { Provider } from "@/components/ui/provider";

const inter = Inter({ subsets: ["latin"] });

export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const { isEnabled } = await draftMode();

  return (
    <html lang="en" suppressHydrationWarning>
      <body className={inter.className}>
        {/* {isEnabled && <PreviewNotice />}
        <Navigation /> */}
        <Provider>{children}</Provider>
      </body>
    </html>
  );
}
```
Note: I'm using Next.js `cms-wordpress` template. However, I disabled some of the additional components that come with that template in the `RootLayout` as you see in the code above.

I attempted using and not using `--turbopack`, but it is still showing the same error.

Now, it seems to render Chakra components without issues despite showing that error.

Finally, I attempted downgrading Next.js to the latest stable 14 version, but the issue remains.

Additional Note: This is my first time using Chakra v3. I had set up Chakra v2 with Next.js 14.1.4 on another project, and I didn't come across this issue.

I'd like to use Chakra v3, but I'm not sure if that hydration issue is going to cause later issues in production.

**Comment by thejoyfuldev** [comment]
I was able to get rid of the error and the need for `suppressHydrationWarning` by ensuring SSR was turned off for `ChakraProvider`

```tsx
// src/components/ui/Provider/provider.tsx
'use client'

import {
  ColorModeProvider,
  type ColorModeProviderProps,
} from '@/components/ui/color-mode'
import { Toaster } from '@/components/ui/toaster'
import { ChakraProvider, defaultSystem } from '@chakra-ui/react'

export function Provider(props: ColorModeProviderProps) {
  return (
    <ChakraProvider value={defaultSystem}>
      <ColorModeProvider {...props} />
      <Toaster />
    </ChakraProvider>
  )
}
```

```ts
// src/components/ui/Provider/index.ts
'use client'

import dynamic from 'next/dynamic'

export const Provider = dynamic(
  () => import('./provider').then((mod) => mod.Provider),
  {
    ssr: false,
  },
)
```

```tsx
// src/app/layout.tsx
import { Provider as ChakraProvider } from '@/components/ui/Provider'
import { Outfit } from 'next/font/google'

const outfitSans = Outfit({
  subsets: ['latin'],
  variable: '--font-outfit',
})

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode
}>) {
  return (
    <html lang='en' className={outfitSans.variable}>
      <body>
        <ChakraProvider>{children}</ChakraProvider>
      </body>
    </html>
  )
}
```

I did notice that `ChakraProvider` must be wrapped around `{children}` and attempting to move it higher in the tree results in other errors. But, this resolves the hydration warning errors.

**Comment by ducnguyenhong** [comment]
I am using NextJS 15 (app router) and Chakra UI version 3.2.0 and also got the same error if using **turbopack** mode from NextJS.

You can quickly fix the error (in the unorthodox way from some comments shared above)
- disable turbopack mode in next dev
- or use dynamic import Provider

Turbopack is promoted by NextJS to help the development process in development environment (next dev). And they plan to use it in production environment (next build).

As a lover of both NextJS and ChakraUI, I really want to have this combination. We need to wait for the developers to fix the bug in the future.

**Comment by kanehooper** [comment]
As an immediate solution and using @ducnguyenhong advice of dynamic imports. Here is how I structure my layout in Next v15 and Chakra v3.

``` layout.tsx
// layout.tsx
import type { Metadata } from 'next'
import localFont from 'next/font/local'
import './globals.css'
import { ReactNode } from 'react'
import ClientProvider from './ClientProvider'

const geistSans = localFont({
  src: './fonts/GeistVF.woff',
  variable: '--font-geist-sans',
  weight: '100 900',
})
const geistMono = localFont({
  src: './fonts/GeistMonoVF.woff',
  variable: '--font-geist-mono',
  weight: '100 900',
})

export const metadata: Metadata = {
  title: 'Create Next App',
  description: 'Generated by create next app',
}

export default function RootLayout({ children }: { children: ReactNode }) {
  return (
    <html lang="en" suppressHydrationWarning>
      <body className={`${geistSans.variable} ${geistMono.variable}`}>
        <ClientProvider>{children}</ClientProvider>
      </body>
    </html>
  )
}
```

The dynamic import needs to be executed in a client component:

``` ClientProvider.tsx
// ClientProvider.tsx
'use client'

import dynamic from 'next/dynamic'

const Provider = dynamic(
  () => import('@/components/ui/provider').then((mod) => mod.Provider),
  {
    ssr: false,
  }
)

export default function ClientProvider({
  children,
}: {
  children: React.ReactNode
}) {
  return <Provider>{children}</Provider>
}
```
It's a bit ugly, but I no longer get the Hydration errors in the console.

**Reply by lucaslosi** [reply]
important to know that this solution slows pagespeed performance

**Comment by gcb** [comment]
i'm always getting errors because of darkmode :disappointed:  already went over the next-theme workaround docs a few times.

```
// layout.tsx
...
	return (
		<html suppressHydrationWarning lang="en"><body><Provider>
...
```
(Provider is the ColorProvider created by the chakraui cli snippet thing)

and error:
```
Hydration failed because the server rendered HTML didn't match the client. As a result this tree will be regenerated on the client. This can happen if a SSR-ed Client Component used:
  ...
    <div className="chakra-con..." ref={null}>
      <OuterLayoutRouter parallelRouterKey="children" template={<RenderFromTemplateContext>} notFound={[...]} ...>
        <RenderFromTemplateContext>
          <ScrollAndFocusHandler segmentPath={[...]}>
            <InnerScrollAndFocusHandler segmentPath={[...]} focusAndScrollRef={{apply:false, ...}}>
              <ErrorBoundary errorComponent={undefined} errorStyles={undefined} errorScripts={undefined}>
                <LoadingBoundary loading={null}>
                  <HTTPAccessFallbackBoundary notFound={[...]} forbidden={[...]} unauthorized={[...]}>
                    <HTTPAccessFallbackErrorBoundary pathname="/login" notFound={[...]} forbidden={[...]} ...>
                      <RedirectBoundary>
                        <RedirectErrorBoundary router={{...}}>
                          <InnerLayoutRouter url="/login" tree={[...]} cacheNode={{lazyData:null, ...}} ...>
                            <RootLayout>
                              <html
                                lang="en"
-                               className="light"
-                               style={{color-scheme:"light"}}
                              >
```

**Comment by saltnpixels** [comment]
I am having the same issue. I am not using turbopack, I made sure the provider was inside body....
The onyl thing that seems to be working so far is manually creating a Emotion Cache Provider. Cause emotion and nextjs don't seem to play nicely. I thought chakra took care of the emotion stuff but maybe not?

in my nextjs.config I added this and so far it seems to help:
```
import type { NextConfig } from 'next';

const nextConfig: NextConfig = {
  /* config options here */
  experimental: {
    optimizePackageImports: ['@chakra-ui/react'],
  },
// the lines to add
  compiler: {
    emotion: true,
  },
};

export default nextConfig;
```

**Reply by gcb** [reply]
For the case of the comment above, i ended up moving all the darkmode buttons to a nested 'use client' and it somewhat solved the hydration errors... but now it is *extremely slow* to actually hydratate the page. on slow mobiles/emulators, it takes a good second (or two!) to be able to click things. and the darkmode toggle button itself still needs the `suppressHydrationWarning` attribute.

**Reply by saltnpixels** [reply]
my compiler emotion true works but breaks any css prop if you use variables :/ 
`css={{color: red.500}}` will break so... nevermind.

**Comment by bertho-zero** [comment]
I managed to get rid of my hydration issues in Next 16 by adding the following configuration to my `next.config.js`:

```js
transpilePackages: [
  '@chakra-ui/react',
  '@emotion/*',
  '@zag-js/*',
],
```

**Reply by nickvisut** [reply]
tried this out and seem to be getting same hydration error as OP with Next.js 16, unfortunately.

**Comment by lucasishuman** [comment]
Now that Turbopack is the default bundler for Next.js projects, as of v.16, it would be great for the Chakra team to take another look at this issue and have a better "official" way of handling it. (while you _can_ opt out of Turbopack, for now, the future of Next is Turbopack)

Upgrading to Next.js 16 guide + Turbopack ~
https://nextjs.org/docs/app/guides/upgrading/version-16#turbopack-by-default

**Comment by thejoyfuldev** [comment]
Versions:
Next.js: 16.2.1
Chakra: 3.34.0
Emotion: 11.14

Here is an updated solution that also fixes the latest error since adding the Date/Calendar component. This uses the new `useServerInsertedHTML` hook to inject rules before any content might use theme.

This is adapted from both sources below:

Emotion issue addressing this from two years ago.
https://github.com/emotion-js/emotion/issues/2928#issuecomment-1293012737

Next.js official doc's about CSS-in-JS
https://nextjs.org/docs/app/guides/css-in-js

```ts
// next.config.ts
import type { NextConfig } from "next";

const nextConfig: NextConfig = {
  reactCompiler: true,
  experimental: {
    optimizePackageImports: ["@chakra-ui/react"],
  },
  compiler: {
    emotion: true,
  },
};

export default nextConfig;
```

```tsx
// EmotionRegistry.tsx
"use client";

import createCache from "@emotion/cache";
import { CacheProvider } from "@emotion/react";
import { useServerInsertedHTML } from "next/navigation";
import type React from "react";
import { useState } from "react";

export default function EmotionRegistry({
  children,
}: {
  children: React.ReactNode;
}) {
  const [{ cache, flush }] = useState(() => {
    const cache = createCache({ key: "css" });
    cache.compat = true;
    const prevInsert = cache.insert;
    let inserted: string[] = [];
    cache.insert = (...args) => {
      const serialized = args[1];
      if (cache.inserted[serialized.name] === undefined) {
        inserted.push(serialized.name);
      }
      return prevInsert(...args);
    };
    const flush = () => {
      const prevInserted = inserted;
      inserted = [];
      return prevInserted;
    };
    return { cache, flush };
  });

  useServerInsertedHTML(() => {
    const names = flush();
    if (names.length === 0) return null;
    let styles = "";
    for (const name of names) {
      styles += cache.inserted[name];
    }
    return (
      <style
        key={cache.key}
        data-emotion={`${cache.key} ${names.join(" ")}`}
        // biome-ignore lint/security/noDangerouslySetInnerHtml: Avoid hydration errors
        dangerouslySetInnerHTML={{ __html: styles }}
      />
    );
  });

  return <CacheProvider value={cache}>{children}</CacheProvider>;
}
```

```tsx
// app/layout.tsx
export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en" className={fontSans.variable} suppressHydrationWarning>
      <body>
        <EmotionRegistry>
          <Provider>
            {children}
            <Toaster />
          </Provider>
        </EmotionRegistry>
      </body>
    </html>
  );
}
```

**Comment by segunadebayo** [comment]
Closing as stale. Please start a new discussion if this is still relevant on the latest version.
